src/tensors/dbcsr_tensor_api_c.F

```diff
@@ -0,0 +1,1231 @@
+      INTEGER, DIMENSION(:, :), ALLOCATABLE              :: bounds_1
+      INTEGER, DIMENSION(:, :), ALLOCATABLE              :: bounds_2
+      INTEGER, DIMENSION(:, :), ALLOCATABLE              :: bounds_3
+      LOGICAL, ALLOCATABLE                               :: optimize_dist
+      TYPE(dbcsr_t_pgrid_type), POINTER                  :: pgrid_opt_1 => null(), &
+                                               pgrid_opt_2 => null(), pgrid_opt_3 => null()
+
+      REAL(KIND=real_8), ALLOCATABLE                     :: filter_eps
+      INTEGER(KIND=int_8), ALLOCATABLE                   :: flop
+      LOGICAL, ALLOCATABLE                               :: move_data
+      LOGICAL, ALLOCATABLE                               :: retain_sparsity
+      INTEGER, ALLOCATABLE                               :: unit_nr
+      LOGICAL, ALLOCATABLE                               :: log_verbose
+
+      CALL c_f_pointer(c_tensor_1, tensor_1)
+      CALL c_f_pointer(c_tensor_2, tensor_2)
+      CALL c_f_pointer(c_tensor_3, tensor_3)
+
+      contract_1 => c_contract_1
+      contract_2 => c_contract_2
+      map_1 => c_map_1
+      map_2 => c_map_2
+      notcontract_1 => c_notcontract_1
+      notcontract_2 => c_notcontract_2
+
+#:for dim in range(1,4)
+      IF (PRESENT(c_bounds_${dim}$)) THEN
+         ALLOCATE(bounds_${dim}$, SOURCE = c_bounds_${dim}$)
+         bounds_${dim}$ = bounds_${dim}$ + 1
+      ENDIF
+      ALLOCATE (pgrid_opt_${dim}$)
+#:endfor
+
+#:set list = ['optimize_dist', 'filter_eps', 'flop', 'move_data', 'retain_sparsity', 'unit_nr', 'log_verbose']
+#:for var in list
+      IF (PRESENT(c_${var}$)) ${var}$ = c_${var}$
+#:endfor
+
+      CALL dbcsr_t_contract(dbcsr_scalar(c_alpha), tensor_1, tensor_2, dbcsr_scalar(c_beta), &
+                            tensor_3, &
+                            contract_1 + 1, notcontract_1 + 1, &
+                            contract_2 + 1, notcontract_2 + 1, &
+                            map_1 + 1, map_2 + 1, &
+                            bounds_1, bounds_2, bounds_3, &
+                            optimize_dist, pgrid_opt_1, pgrid_opt_2, pgrid_opt_3, &
+                            filter_eps, flop, move_data, retain_sparsity, unit_nr, log_verbose)
+
+#:for dim in range(1,4)
+      IF (PRESENT(c_pgrid_opt_${dim}$)) c_pgrid_opt_${dim}$ = c_loc(pgrid_opt_${dim}$)
+#:endfor
+
+   END SUBROUTINE
+
+#:endfor
+
+   FUNCTION c_ndims_tensor(c_tensor) BIND(C, name="c_ndims_tensor")
+      TYPE(c_ptr), INTENT(IN), VALUE                  :: c_tensor
+      TYPE(dbcsr_t_type), POINTER                     :: tensor
+      INTEGER(kind=c_int)                             :: c_ndims_tensor
+
+      CALL c_f_pointer(c_tensor, tensor)
+
+      c_ndims_tensor = tensor%nd_index%ndim_nd
+   END FUNCTION
+
+#:for dparam, dtype, dsuffix, basetype, cftype in cf_dtype_float_list
+#:for ndim in ndims
+
+   SUBROUTINE c_dbcsr_t_allocate_and_get_${ndim}$d_block_${dsuffix}$ (c_tensor, tensor_dim, c_ind, c_block, c_found) &
```

Comment:
If all allocations should be controlled in C, we should not expose this routine.

I'll write a function to more easily get block sizes.

Reply:
Ok, I will adapt it later then.